SQRLDotNetClientUI/App.xaml.cs

```diff
@@ -49,6 +67,154 @@ public class App : Application
+            SQRL.CPS.CPSRequestReceived += CPSRequestReceived;
+        }
+
+        /// <summary>
+        /// Handles data coming in via Inter-Process-Communication.
+        /// </summary>
+        /// <param name="ipcDataString">The data received by the IPC server. This sould
+        /// either be a valid "sqrl://" URL to initiate an authentication operation, or 
+        /// the magic wakeup keyword if we only want to show the main window.</param>
+        public void HandleIPC(string ipcDataString)
+        {
+            if (ipcDataString == MagicWakeupString)
+            {
+                Log.Information("Showing main screen via IPC wakeup");
+                ShowMainScreen();
+            }
+            else
+            {
+                Log.Information("Try handling auth request via IPC");
+                HandleAuthRequest(ipcDataString, "IPC");
+            }
+        }
+
+        /// <summary>
+        /// This event handler gets called if the CPS server receives a new CPS request.
+        /// </summary>
+        private void CPSRequestReceived(object sender, CPSRequestReceivedEventArgs e)
+        {
+            // We're not interested in "gif" queries
+            if (e.Context.Request.RawUrl.EndsWith(".gif")) return;
+
+            // Extract and base64-decode the sqrl:// URL from the cps request
+            string data = e.Context.Request.Url.AbsolutePath.Substring(1);
+            string sqrlUrl = Encoding.UTF8.GetString(Sodium.Utilities.Base64ToBinary(
+                data, "", Sodium.Utilities.Base64Variant.UrlSafeNoPadding));
+
+            Log.Information("Try handling auth request via CPS");
+            HandleAuthRequest(sqrlUrl, "CPS");
+        }
+
+        /// <summary>
+        /// Checks if an auth request was already performed for the nut embedded within
+        /// <paramref name="sqrlUrl"/>, and if not, fires up the authentication screen.
+        /// Any auth request for a nut which was already handled will be ignored.
+        /// This handles the "race condition" that arises in our new approach where both
+        /// CPS and a sqrl:// scheme invocation can trigger an authentication popup.
+        /// </summary>
+        /// <param name="sqrlUrl">The "sqrl://" authentication URL.</param>
+        /// <param name="source">A string representing the caller of the method. Used for logging.</param>
+        private void HandleAuthRequest(string sqrlUrl, string source)
+        {
+            try
+            {
+                Uri uri = new Uri(sqrlUrl);
+                string nut = HttpUtility.ParseQueryString(uri.Query).Get("nut");
+                if (string.IsNullOrEmpty(nut)) return;
+
+                byte[] hashedNut = Sodium.CryptoHash.Sha256(nut);
+                bool isFirstAuthForThisNut = false;
+
+                Log.Information($"Locking last nut hash for {source} from Thread id {Thread.CurrentThread.ManagedThreadId}");
+                lock (this._priorNutHashLockObject)
```

Comment:
Very nice love this approach nice and clean and whomever gets there first wins #Awesome ++